keg_auth/libs/authenticators.py

```diff
@@ -206,6 +210,61 @@ def post(self, *args, **kwargs):
         self.assign_template_vars(form)
 
 
+class AttemptLimitMixin(object):
+    @property
+    def attempt_ent(self):
+        return flask.current_app.auth_manager.entity_registry.attempt_cls
+
+    def should_limit_attempts(self):
+        try:
+            return self.attempt_ent is not None
+        except RegistryError:
+            return False
+
+    def is_attempt_blocked(self, user):
```

Comment:
I'm not sure the logic on this is quite right, and I don't see any tests to verify it works as expected.

Shouldn't you be getting the number of attempts within the previous `get_attempt_timespan()`, and then basic the count just on that? I'm not sure what the purpose of starting at the `lockout_period` is supposed to be.

Let's say we want to lock a view down for 5 minutes if a user tries to access it more than 3 times within a 15 minute window. I'm not sure this code will allow for that to happen.

Alternatively, let's say we want to lock a view down for 2 hours if a user tries to access it more than 5 times within a 10 minute window. The user accesses the view once every 10 minutes for an hour, and then tries 20 times within the next 5 minutes. Since we're going back by the `lockout period` and then moving forward by the `attempt timespan`, I don't think those most-recent attempts would be counted.

Reply:
There are some things wrong here:
- Generally speaking, I think we should only lock after a number of consecutive unsuccessful attempts within the time period.
  - Which means, don't penalize the user for logging in successfully sometime within that block of attempts. If I'm locking after three bad attempts, and I log 2 unsuccessful before the user logs in correctly, the next bad attempt shouldn't lock.
  - That will affect how generic this logic is between logins and resets, but that can be factored out for how you grab your last "milestone" event (last successful login, or framing your window of reset attempts)
- I share TJ's concerns about the edge cases here, so it's worth another look and perhaps some additional "scenario" tests for some of the what-ifs. FWIW, if you search RRS for `login_id_is_locked` there is an alternate solution that's just missing the attempt timespan (you'd want to pull the date math out of the query there though)